[WIP] RabbitMQ native queue length

tmasternak opened the pull request.

bording commented:
> Am I handling connection recoverability in a correct way i.e. what happens if the connections goes down?

With automatic recovery enabled, it will reconnect if it loses the connection to the broker. However, if you try to use the channel and connection while in the reconnecting state, they'll throw exceptions.
 I'd recommend using the `ConnectionShutdown` and `RecoverySucceeded` Connection events to detect these states and stop/start your loop.

`ConnectionShutdown` has `ShutdownEventArgs` which can be used to determine why the connection is closing. You can see an example of that in the RabbitMQ transport message pump.

> I did not verify what happens with DefineQueuePasive if the queue for some reason does not exists. Do you know if the channel still usable or should I recreate it?

If the queue does not exist, you get an channel exception and the channel is closed. You'll have to catch an exception and create a new channel.


BTW, there is actually a `MessageCount` API on the channel that you could use instead of `QueueDeclarePassive`. At the protocol level doing the same thing as `QueueDeclarePassive`, but so it's basically a convenience method.

bording commented on src/ServiceControl.Transports.RabbitMQ/QueueLengthProvider.cs:
```diff
@@ -0,0 +1,142 @@
+            var factory = new ConnectionFactory
+            {
+                Uri = connectionString,
+                AutomaticRecoveryEnabled = true,
```
`AutomaticRecoveryEnabled` is true by default for the current client, so you don't need to specify it.

bording commented on src/ServiceControl.Transports.RabbitMQ/QueueLengthProvider.cs:
```diff
@@ -0,0 +1,142 @@
+            {
+                Uri = connectionString,
+                AutomaticRecoveryEnabled = true,
+                TopologyRecoveryEnabled = true
```
This one is true by default as well, but I'd recommend setting it to false instead. You don't actually have anything here to recover, so there's no reason to have it enabled.

bording commented on src/ServiceControl.Transports.RabbitMQ/QueueLengthProvider.cs:
```diff
@@ -0,0 +1,142 @@
+
+        void FetchQueueLengths(IModel model)
+        {
+            //TODO: check what happens when non existing queue is querried
```
As mentioned in comment, it throws an exception and the channel is closed.

bording commented on src/ServiceControl.Transports.RabbitMQ/QueueLengthProvider.cs:
```diff
@@ -0,0 +1,142 @@
+            {
+                try
+                {
+                    var declareOK = model.QueueDeclarePassive(endpointQueuePair.Value);
```
As mentioned already, you can use `model.MessageCount` instead.

bording commented on src/ServiceControl.Transports.RabbitMQ/QueueLengthProvider.cs:
```diff
@@ -0,0 +1,142 @@
+        {
+            var factory = new ConnectionFactory
+            {
+                Uri = connectionString,
```
What sort of connection string are you expecting here? The format described in https://docs.particular.net/transports/rabbitmq/connection-settings is something we created, and is not what the client is expecting.

The `Uri` should be something like `amqp://rabbit.mshome.net:5672`. 

For consistency, it might make sense to bring in the code from the transport to support the same type of connection string.

tmasternak commented:
@bording thanks for the input!

I've added the connection string parsing logic (taken from the transport) and moved the connection, model failure handling to a separate class `QueryExecutor`. I did not go with the callback - checking the status flag looks good enough to me. Thoughts?

mikeminutillo commented on src/ServiceControl.Transports.RabbitMQ/QueueLengthProvider.cs:
```diff
@@ -0,0 +1,177 @@
+                        new RawMessage.Entry
+                        {
+                            DateTicks = nowTicks,
+                            Value = sizes.TryGetValue(endpointQueuePair.Value, out var size) ? size : 0
```
Is it correct to report 0 where we have no measurement?

tmasternak replied:
This is a good point.

mikeminutillo commented on src/ServiceControl.Transports.RabbitMQ/QueueLengthProvider.cs:
```diff
@@ -0,0 +1,177 @@
+                }
+            }
+
+            TimeSpan ReconnectionDelay = TimeSpan.FromSeconds(5);
```
This is _forever_. What if the reconnection happens just after we checked the flag?

tmasternak replied:
This is based on the fact that 5 sec is the default [timeout](https://www.rabbitmq.com/dotnet-api-guide.html#connection-recovery) for the built-in rabbitmq client library recoverability.

tmasternak replied:
Actually the default used by RabbitMQ client is 10 sec. See @bording  [comment](https://github.com/Particular/ServiceControl.Monitoring/pull/133#discussion_r193945454)

bording replied:
10 seconds is the default that the transport uses when creating a `ConnectionFactory`, but the client's default value if the `NetworkRecoveryInterval` property isn't modified is 5 seconds.

tmasternak commented on src/ServiceControl.Transports.RabbitMQ/ConnectionConfiguration.cs:
```diff
@@ -0,0 +1,194 @@
+                invalidOptionsMessage.AppendLine("Invalid connection string. 'host' value must be supplied. e.g: \"host=myServer\"");
+            }
+
+            var nsbVersion = FileVersionInfo.GetVersionInfo(typeof(Endpoint).Assembly.Location);
```
Transport specific packages already have (or are predicted to have) transport reference and as a result core reference so I think this is unavoidable.

tmasternak commented:
@bording can you have a look?
